File: src/main/java/com/example/dq/util/CryptoUtil.java
package com.example.dq.util;

import com.example.dq.config.DqProperties;
import org.springframework.stereotype.Component;

import javax.crypto.Cipher;
import javax.crypto.spec.GCMParameterSpec;
import javax.crypto.spec.SecretKeySpec;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.SecureRandom;
import java.util.Base64;

/** 数据源密码 AES-GCM 对称加密 */
@Component
public class CryptoUtil {

    private final SecretKeySpec key;
    private final SecureRandom random = new SecureRandom();

    public CryptoUtil(DqProperties props) throws Exception {
        byte[] raw = MessageDigest.getInstance("SHA-256")
                .digest(props.getSecurity().getSecret().getBytes(StandardCharsets.UTF_8));
        this.key = new SecretKeySpec(raw, "AES");
    }

    public String encrypt(String plain) {
        if (plain == null || plain.isEmpty()) {
            return plain;
        }
        try {
            byte[] iv = new byte[12];
            random.nextBytes(iv);
            Cipher cipher = Cipher.getInstance("AES/GCM/NoPadding");
            cipher.init(Cipher.ENCRYPT_MODE, key, new GCMParameterSpec(128, iv));
            byte[] ct = cipher.doFinal(plain.getBytes(StandardCharsets.UTF_8));
            byte[] out = new byte[iv.length + ct.length];
            System.arraycopy(iv, 0, out, 0, iv.length);
            System.arraycopy(ct, 0, out, iv.length, ct.length);
            return Base64.getEncoder().encodeToString(out);
        } catch (Exception e) {
            throw new IllegalStateException("加密失败", e);
        }
    }

    public String decrypt(String enc) {
        if (enc == null || enc.isEmpty()) {
            return enc;
        }
        try {
            byte[] all = Base64.getDecoder().decode(enc);
            byte[] iv = new byte[12];
            System.arraycopy(all, 0, iv, 0, 12);
            Cipher cipher = Cipher.getInstance("AES/GCM/NoPadding");
            cipher.init(Cipher.DECRYPT_MODE, key, new GCMParameterSpec(128, iv));
            return new String(cipher.doFinal(all, 12, all.length - 12), StandardCharsets.UTF_8);
        } catch (Exception e) {
            throw new IllegalStateException("解密失败", e);
        }
    }
}

File: dq-desktop/src/main/kotlin/com/example/dq/util/CryptoUtil.kt
package com.example.dq.util

import com.example.dq.config.AppConfig
import java.nio.charset.StandardCharsets
import java.security.MessageDigest
import java.security.SecureRandom
import java.util.Base64
import javax.crypto.Cipher
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec

/** 数据源密码 AES-GCM 对称加密 */
class CryptoUtil(config: AppConfig) {

    private val key: SecretKeySpec
    private val random = SecureRandom()

    init {
        val raw = MessageDigest.getInstance("SHA-256")
            .digest(config.securitySecret.toByteArray(StandardCharsets.UTF_8))
        key = SecretKeySpec(raw, "AES")
    }

    fun encrypt(plain: String?): String? {
        if (plain.isNullOrEmpty()) {
            return plain
        }
        try {
            val iv = ByteArray(12)
            random.nextBytes(iv)
            val cipher = Cipher.getInstance("AES/GCM/NoPadding")
            cipher.init(Cipher.ENCRYPT_MODE, key, GCMParameterSpec(128, iv))
            val ct = cipher.doFinal(plain.toByteArray(StandardCharsets.UTF_8))
            val out = ByteArray(iv.size + ct.size)
            System.arraycopy(iv, 0, out, 0, iv.size)
            System.arraycopy(ct, 0, out, iv.size, ct.size)
            return Base64.getEncoder().encodeToString(out)
        } catch (e: Exception) {
            throw IllegalStateException("加密失败", e)
        }
    }

    fun decrypt(enc: String?): String? {
        if (enc.isNullOrEmpty()) {
            return enc
        }
        try {
            val all = Base64.getDecoder().decode(enc)
            val iv = ByteArray(12)
            System.arraycopy(all, 0, iv, 0, 12)
            val cipher = Cipher.getInstance("AES/GCM/NoPadding")
            cipher.init(Cipher.DECRYPT_MODE, key, GCMParameterSpec(128, iv))
            return String(cipher.doFinal(all, 12, all.size - 12), StandardCharsets.UTF_8)
        } catch (e: Exception) {
            throw IllegalStateException("解密失败", e)
        }
    }
}
